tests/cpp/test_multidevice_host_ir.cpp

```diff
@@ -408,6 +408,67 @@ TEST_F(OverlapDistributedMatmulTest, AG_matmul) {
   EXPECT_TRUE(torch::allclose(tc_ref, tc, 1e-2, 1e-2));
 }
 
+TEST_F(OverlapDistributedMatmulTest, AG_linear) {
+  constexpr int64_t M = 32768;
+  constexpr int64_t K = 32768;
+  constexpr int64_t N = 1024;
+  constexpr int64_t S = 8;
+  const int64_t D = communicator_->size();
+  if (M % (D * S) != 0) {
+    GTEST_SKIP() << "M must be a multiple of D * S, but got M = " << M
+                 << ", D = " << D << ", S = " << S;
+  }
+
+  auto fusion = std::make_unique<Fusion>();
+  FusionGuard fg(fusion.get());
+
+  TensorView* in = makeContigTensor(4); //[S, DIDx(D), M/(S*D), K]
+  TensorView* weight = makeContigTensor(2); //[N, K]
+  TensorView* bias = makeContigTensor(1); //[N]
+  TensorView* out = linear(in, weight, bias); //[S, D, M/(S*D), N]
+
+  fusion->addInput(in);
+  fusion->addInput(weight);
+  fusion->addInput(bias);
+  fusion->addOutput(out);
+
+  auto mesh = DeviceMesh::createForNumDevices(D);
+  in->setDeviceMesh(mesh);
+  weight->setDeviceMesh(mesh);
+  bias->setDeviceMesh(mesh);
+  out->setDeviceMesh(mesh);
+
+  in->axis(1)->parallelize(ParallelType::DIDx);
+  out->axis(0)->parallelize(ParallelType::Stream);
+
+  MultiDeviceExecutor executor(std::move(fusion), *communicator_);
+
+  auto tensor_options =
+      at::TensorOptions().dtype(at::kFloat).device(communicator_->device());
+  at::Tensor in_at_unsharded =
+      at::randn({S, D, M / (S * D), K}, tensor_options);
+  at::Tensor in_at = in_at_unsharded.slice(
+      1, communicator_->deviceId(), communicator_->deviceId() + 1);
+  at::Tensor weight_at = at::randn({N, K}, tensor_options);
+  at::Tensor bias_at = at::randn({N}, tensor_options);
+  at::Tensor out_ref = at::linear(in_at_unsharded, weight_at, bias_at);
+
+  std::vector<c10::IValue> inputs = {in_at, weight_at, bias_at};
+  at::Tensor out_at;
+
+  constexpr int64_t kNumberOfIterations = 20;
+  constexpr int64_t kNumberOfWarmupIterations = 5;
+  for (auto i : c10::irange(kNumberOfIterations)) {
+    if (i == kNumberOfWarmupIterations) {
+      cudaProfilerStart();
+    }
+    out_at = executor.runWithInput(inputs).at(0);
+  }
+  cudaProfilerStop();
```

Comment:
Should there be a device sync before this? https://docs.nvidia.com/cuda/profiler-users-guide/#flush-profile-data

Reply:
you're right, thanks!